### ISSUE TITLE
verify Clang's fortified string routines are called in the kernel

### ISSUE THREAD

**Issue by nickdesaulniers** [ClangBuiltLinux/linux#1002]
forking #979 which focuses on regressions. cc @kees @gburgessiv @m-gupta 
(If we're happy with a new bug, please close #979 as fixed, otherwise close this as a dup of #979).

Kees mentions that the runtime implementations of fortified string routines in the kernel may never have been called in a clang built kernel, irrespective of regressions in #979 ? If that's so, we should figure out what we need to fix and where to leverage this technology.  It's probably possible to write a kselftest to help prevent support from regressing.

**Comment by gburgessiv** [comment]
Thanks for filing! I hear no discomfort with #979's resolution, so I'll close it.

> Kees mentions that the runtime implementations of fortified string routines in the kernel may never have been called in a clang built kernel, irrespective of regressions in #979 ?

This depends on the implementation of the routine in question, but yeah, clang won't generate IR for `extern inline gnu_inline` functions that appear trivially recursive in a particular way. This turns at least some of the kernel's FORTIFY into a nop.

There's some discussion of approaches we can try on https://github.com/ClangBuiltLinux/linux/pull/980#discussion_r405824303. Kees noted that automated testing would be really nice to have here before we start trying to thread this particular needle. I agree with that. :)

**Comment by nathanchance** [comment]
https://lore.kernel.org/lkml/20200506001453.764332-1-Jason@zx2c4.com/

**Comment by nathanchance** [comment]
@kees had mentioned recently that he was still looking into this?

**Comment by kees** [comment]
Yup, currently investigating this as a part of my memcpy work.

**Comment by kees** [comment]
This is blocked on an LLVM bug: https://llvm.org/pr41459

**Comment by kees** [comment]
And blocked on this LLVM bug: https://llvm.org/pr50322

**Comment by nickdesaulniers** [comment]
jotting down notes from trying to run LKDTM:
```sh
/ # mount -t sys /sys
/ # mount -t debugfs debugfs /sys/kernel/debug
/ # cat /sys/kernel/debug/provoke-crash/DIRECT
/ # echo FORTIFIED_OBJECT | cat > /sys/kernel/debug/provoke-crash/DIRECT
```

**Comment by nathanchance** [comment]
This is fixed with https://git.kernel.org/linus/281d0c962752fb40866dd8d4cade68656f34bd1f.

#1583 is tracking the `!X86_32` in `CONFIG_FORTIFY_SOURCE`'s entry.
